src/Build/Instance/ProjectInstance.cs

```diff
@@ -326,6 +326,59 @@ public ProjectInstance(ProjectRootElement xml, IDictionary<string, string> globa
         {
         }
 
+        /// <summary>
+        /// Creates a ProjectInstance from an external created <see cref="Project"/>.
+        /// Properties and items are cloned immediately and only the instance data is stored.
+        /// </summary>
+        public ProjectInstance(Project project, ProjectInstanceSettings settings)
```

Comment:
How is this different from the other [ProjectInstance(Evaluation.Project.Data)](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Instance/ProjectInstance.cs#L451) constructor which is called from [Project.CreateProjectInstance](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Definition/Project.cs#L1333)? They seem the same to me, so I'd rather not add this additional constructor (unless I'm missing something).

Reply:
Ah, okay, let me rephrase it to ensure that I got it right: your scenario is to create a `ProjectInstance` object residing in Process_1 based on a `Project` object residing in Process_2. You want to avoid marshaling the `Project.Data` object from Process_2 to Process_1 and instead have the `ProjectInstance` constructor from Process_1 do remote calls for each `Project` data field (one "rpc" call for properties, one for items, one for targets, etc).

Yea, in this case reading the data directly from the remote Project object is more straighforward than manually coding up a stub/proxy for `Project.Data` as well. Assuming the two processes always reside on the same machine, otherwise the rpc overhead might make it worth it to serialize the entire Project.Data over one rpc call.

However, I'd just delete the old `ProjectInstance(Evaluation.Project.Data)` constructor. It's just code duplication with your new constructor.

Reply:
Yes, that's a good summary. The old `ProjectInstance(Evaluation.Project.Data)` constructor still will get used as-is in the non cached mode in VS where MSBuild gets called for actual evaluation. I would like to keep it that way for feature isolation reasons, atleast until the cache mode feature which will initially go in behind a feature flag gets more test coverage.